[TECH] Amérliorer la gestion de l'authentification dans Hapi (PIX-2408).

aceol opened the pull request:
## :unicorn: Problème
Avec la mise en place de l'APIM/Livret scolaire, il y a plusieurs possibilité pour acceder à l'API Pix:
- user token
- application token
- pas de token

Cette partie du code n'est pas claire est melange plusieurs concepts. Elle nécessite un refacto


## :robot: Solution
Mettre en place un scheme unique (jwt-schema), et utiliser les options pour passer une handler de validation selon le besoin (user/application)

## :rainbow: Remarques
Plutôt que de renvoyer des erreurs domain custom, on utilise directement Boom. Le message renvoyé est peut donc être different en cas d'erreur d'authentification, mais cela ne devrait pas avoir d'impact.

On devrait pouvoir remplacer l'utilisation de `extractUserIdFromRequest` des controlleurs

## :100: Pour tester
- [ ] Le passage du token "user" devrait fonctionner correctement
- [ ] Le passage du token "application" devrait fonctionner correctement
- [ ] Les endpoints avec un auth en `optional` peuvent être accédé avec ou sans authentification

aceol committed:
[SR] Remove useless stub

pix-service commented:
I'm deploying this PR to these urls:

- App: https://app-pr2693.review.pix.fr
- Orga: https://orga-pr2693.review.pix.fr
- Certif: https://certif-pr2693.review.pix.fr
- Admin: https://admin-pr2693.review.pix.fr
- API: https://api-pr2693.review.pix.fr/api/

 Please check it out!

aceol committed:
[SR] Remove useless step

aceol committed:
[SR] Check for user instead of isAuthenticated

There can be an authentication (token), but the token is revoked
so it is better to check for the current user instead

aceol committed:
[SR] remove blank space

aceol committed:
[SR] Create user before loading it (hidden error)

aceol committed:
[SR] Create missing userOrgaSettings (hidden error)

aceol committed:
[SR] invalidate session whatever the error is

aceol committed:
[FEATURE] use 2 auth strategy for 1 scheme

MathieuGilet commented on orga/app/services/current-user.js:
```diff
@@ -28,10 +27,7 @@ export default class CurrentUserService extends Service {
-        if ([401, 403].includes(errorCode)) {
-          return this.session.invalidate();
-        }
+        return this.session.invalidate();
```
Il y a une raison à ne pas rendre `null` le prescriber ici (current user dans Pix Orga) comme tu l'as fait pour les autres services ?

chrc replied:
+1 ?

aceol replied:
Non, autant le mettre aussi pour s'assurer de ne pas avoir des données corrompus

MathieuGilet commented on admin/app/services/current-user.js:
```diff
@@ -12,9 +11,8 @@ export default class CurrentUserService extends Service {
-        if (get(error, 'errors[0].code') === 401) {
-          return this.session.invalidate();
-        }
+        this.user = null;
```
Il serait peut-être bien d'appliquer ce fonctionnement aussi à Pix Certif, non ?

chrc commented on mon-pix/app/routes/error.js:
```diff
@@ -6,8 +6,8 @@ export default class ErrorRoute extends Route {
-    const statusCode = get(error, 'errors[0].code');
-    return statusCode === 401;
+    const statusCode = get(error, 'errors[0].status');
+    return statusCode === 401 + '';
```
Question : peut-on utiliser `parseInt(statusCode)` ?

MathieuGilet commented on api/lib/application/pre-response-utils.js:
```diff
@@ -9,6 +9,11 @@ function handleDomainAndHttpErrors(request, h) {
     return errorManager.handle(request, h, response);
   }
 
+  // Ne devrait pas etre necessaire
```
C'est nécessaire pour formater l'erreur Boom renvoyée lorsque la requête ne contient pas de header `Authorization`.

chrc commented on api/tests/unit/application/scorecards/index_test.js:
```diff
@@ -15,9 +14,6 @@ describe('Unit | Router | scorecard-router', () => {
-        h.continue({ credentials: { accessToken: 'jwt.access.token' } });
-      });
       sinon.stub(scorecardController, 'getScorecard').returns('ok');
       startServer();
```
Suggestion : utiliser le helper `HttpTestServer`

aceol replied:
Ah oui c'est top. Je met ça dans un commit à part

aceol committed:
[SR] replace hapi server with HttpTestServer
